delfin/drivers/dell_emc/vmax/client.py

```diff
@@ -287,6 +287,123 @@ def list_volumes(self, storage_id):
+                    'status': status,
+                    'location':
+                        'slot_' +
+                        str(director_info.get('director_slot_number')),
+                    'soft_version': None,
+                    'cpu_info': 'Cores-'
+                                + str(director_info.get('num_of_cores')),
+                    'memory_size': None
+
+                }
+                controller_list.append(controller)
+            return controller_list
+
+        except exception.SSLCertificateFailed:
+            LOG.error('SSL certificate failed when list pools for VMax')
+            raise
+        except Exception as err:
+            msg = "Failed to get controller metrics from VMAX: {}".format(err)
+            LOG.error(msg)
+            raise exception.ControllerNotFound(self.array_id)
+
+    def list_ports(self, storage_id):
+        try:
+            # Get list of Directors
+            directors = self.rest.get_director_list(self.array_id,
+                                                    self.uni_version)
+            switcher = {
+                'A': constants.PortLogicalType.MANAGEMENT,
+                'B': constants.PortLogicalType.SERVICE,
+                'C': constants.PortLogicalType.BACKEND,
+            }
+            port_list = []
+            for director in directors:
+                port_keys = self.rest.get_port_list(
+                    self.array_id, self.uni_version, director)
+                for port_key in port_keys:
+                    port_info = self.rest.get_port(
+                        self.array_id, self.uni_version,
+                        director, port_key['portId'])['symmetrixPort']
+
+                    connection_status = \
+                        constants.PortConnectionStatus.CONNECTED
+                    if port_info.get('port_status',
+                                     '').upper().find('OFF') != -1:
+                        connection_status = \
+                            constants.PortConnectionStatus.DISCONNECTED
+
+                    port_type = constants.PortType.OTHER
+                    if port_info.get('type', '').upper().find('FIBRE') != -1:
+                        port_type = constants.PortType.FC
+                    if port_info.get('type', '').upper().find('ETH') != -1:
+                        port_type = constants.PortType.ETH
+
+                    name = "{0}:{1}".format(port_key['directorId'],
+                                            port_key['portId'])
+
+                    director_emulation = port_key['directorId'][4]
+                    logical_type = switcher.get(
+                        director_emulation, constants.PortLogicalType.OTHER)
+                    if logical_type == constants.PortLogicalType.OTHER:
+                        port_prefix = port_key['directorId'][:2]
+                        if port_prefix in ['FA', 'FE', 'EA', 'EF', 'SE']:
```

Comment:
I guess it is safe to have one more comparison here to map backend directors [DF,DA,DX]-> backend. becasue D-H directors are universal.